examples/step-57/step-57.cc

```diff
@@ -0,0 +1,948 @@
+  // (if the bound of mesh is two, that means the mesh will be refined once). "initial" is the flag to tell us
+  // whether "setup_system" is necessary, and which part, system matrix or right hand side vector, should be assembled.
+  // If we do line search, the right hand side is already assembled while checking residual norm in last iteration.
+  // Therefore, we just assemble the system matrix at current iteration. If the number of meshes is not one, we must
+  // assemble the whole system at the beginning. And this system depends on present solution which is transfered from
+  // the solution on last mesh. The last argument "output_result" is for output.
+
+  template <int dim>
+  void Navier_Stokes_Newton<dim>::newton_iteration(const double tolerance,
+                                                   const unsigned int max_iteration,
+                                                   const unsigned int n_refinements,
+                                                   bool  is_initial_step,
+                                                   bool  output_result)
+  {
+    double current_res;
+    double last_res;
+    bool   first_step = is_initial_step;
+
+    for (unsigned int refinement = 0; refinement < n_refinements+1; ++refinement)
+      {
+        unsigned int outer_iteration = 0;
+        last_res = 1.0;
+        current_res = 1.0;
+        std::cout << "*****************************************" << std::endl;
+        std::cout << "************  refinement = " << refinement << " ************ " << std::endl;
+        std::cout << "viscosity= " << viscosity << std::endl;
+        std::cout << "*****************************************" << std::endl;
+
+        while ((first_step || (current_res > tolerance)) && outer_iteration < max_iteration)
+          {
+            if (first_step)
+              {
+                setup_dof();
+                initialize_system();
+                evaluation_point = present_solution;
+                assemble_NavierStokes_system(first_step, true, true);
+                solve(first_step);
+                present_solution = newton_update;
+                nonzero_constraints.distribute(present_solution);
+                first_step = false;
+                evaluation_point = present_solution;
+                assemble_NavierStokes_system(first_step, false, true);
+                current_res = system_rhs.l2_norm();
+                std::cout << "******************************" << std::endl;
+                std::cout << " The residual of initial guess is " << current_res << std::endl;
+                std::cout << " Initialization complete!  " << std::endl;
+                last_res = current_res;
+              }
+
+            else
+              {
+                evaluation_point = present_solution;
+                if (outer_iteration == 0)
+                  {
+                    assemble_NavierStokes_system(first_step, true, true);
```

Comment:
@tjhei:
> I am pretty sure the matrix is reassembled in each step.

Isn't this line saying that you only do it in the first outer iteration? And outer iteration referring to the Newton iteration? (I might be missing something here, which would not be surprising when looking at 1.7k lines of new code/comments for the first time.)